hw/ip/csrng/doc/_index.md

```diff
@@ -0,0 +1,89 @@
+- The IP interfaces with external entropy sources to obtain any required non-determinstic seed material (entropy) and nonces.
+    - Requires an external entropy source which is compliant with NIST SP 800-90B, and which also satisfies the requirements for a PTG.2 class physical non-deterministic random number generator as defined in AIS31.
+    - Dedicated hardware interface with external entropy source satisfies requirements for `get_entropy_input()` interface as defined in SP 800-90A.
+    - When needed, utilizes the `Block_Cipher_df` derivation function (as defined in SP 800-90A) for assembling seed material.
+      This allows the use of entropy sources which are not full-entropy (less than one bit of entropy per bit generated).
+- Also supports the optional use of personalization strings or other application inputs (e.g. OTP memory values) during instantiation.
+- Assuming a continuously-live entropy source, each instance can also optionally be used as a non-determinstic TRNG (true random number generator, also called a non-deterministic random bit generator or NRBG in SP 800-90C).
+    - In this mode, an instance also meets the requirements laid out for a PTG.3 class RNG, the strongest class laid out in AIS31.
+    - Implementation follows the NRBG "Oversampling Construction" approved by SP 800-90C, to meet both CC and FIPS TRNG constructions.
+- In addition to the approved DRNG mode, any instance can also operate in "Fully Deterministic mode", meaning the seed depends entirely on application inputs or personalization strings.
+    - This provides an approved means of key-derivation in FIPS 140-2.
+
+## Description
+
+Though the recommendations in AIS31 are based around broad functional requirements, the recommendations in SP 800-90 are very prescriptive in nature, outlining the exact constructs needed for approval.
+Thus the interface and implementation are largely driven by these explicit constructs, particularly the CTR_DBRG construct.
+
+
+The CSRNG IP consists of four main components:
+1. An AES primitive
+2. The CTR_DRBG state-machine (CTR_DRBG_FSM) which drives the AES primitive, performing the various encryption sequences proscribed in SP 800-90A.
+These include:
+
+    1. **The Derivation Function:**  
+       Part of the instantiation and reseed routines, this routine assembles the previous seed material (on reseed only), application inputs, and entropy.
+    2. **The Instantiation Routine:**
+       Combines application inputs, external entropy and nonce (more entropy) via the derivation function.
+    3. **The Reseed Routine:**
+       Combines the previous seed material with external entropy to generate a new seed.
+    4. **The Generate Routine:**
+       Generates up to CSRNG_MAX_GENERATE random bits.
+       If called with prediction_resistance_flag, forces a reseed.
+     5. **The Update Routine:**
+        Updates the internal state of the DRBG instance after each generate call.
```

Comment:
"DRBG instance" -> "DRNG instance" (? - I'm not sure - I just noticed it because you use DRNG instance in the next bullet)

Reply:
Changed.  The fact that CC and NIST use different acronyms is pretty frustrating at times.  Though I've changed this here and  clarified the usage in the first paragraph.

`In this document the terms "DRNG" are "TRNG", are used most generally to refer to deterministic or true random number generation functionalities implemented to this specification.`
`However, the terms "DRBG" or "NRBG" are specifically used when respectively referring to SP 800-90A or SP 800-90C requirements.`
`Meanwhile, when addressing requirements which originate from AIS31 we refer to the specific DRG.3 or PTG.3 classes of RNGs.`

Reply:
Fully agree with your frustration!  That clarification is quite helpful.

Reply:
Though it has a typo which I just noticed, and will fix.